### PR TITLE
Support  symmetric encryption

### PR DESCRIPTION
**Pull request by hwupathum**
## Purpose
Support AES/GCM/NoPadding and AES/ECB/PKCS5Padding symmetric encryption methods.

This requires the following config to be added to secret-conf.properties

```
secretRepositories.file.encryptionMode=symmetric
```

## Related issue
- https://github.com/wso2/product-is/issues/20610

### CODE REVIEW

File: src/main/java/org/wso2/securevault/SymmetricCipher.java [hunk 1]
```diff
@@ -59,6 +69,41 @@ public Key getKey(CipherOperationMode operationMode) {
...
+
+        String algorithm = cipherInformation.getAlgorithm();
+        // For AES-GCM mode, initialization happens during encryption or decryption
+        if (!AES_GCM_NO_PADDING.equals(algorithm)) {
+            super.cipherInit(opMode);
+        }
+    }
+
+    /**
+     * Decrypts input cipher bytes
```
Comment (hwupathum) [inline src/main/java/org/wso2/securevault/SymmetricCipher.java:88]: Add to comment that this is only for GCM

File: src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java [hunk 2]
```diff
@@ -46,7 +55,10 @@ public class FileBaseSecretRepository implements SecretRepository {
     private static final String KEY_STORE = "keyStore";
     private static final String DOT = ".";
     private static final String ALGORITHM = "algorithm";
+    private static final String SYMMETRIC = "symmetric";
+    private static final String ENCRYPTION_MODE = "encryptionMode";
     private static final String DEFAULT_ALGORITHM = "RSA";
```
Comment (hwupathum) [inline src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java:60]: Change name

File: src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java [hunk 3]
```diff
@@ -116,8 +133,14 @@ public void init(Properties properties, String id) {
         cipherInformation.setAlgorithm(algorithm);
         cipherInformation.setCipherOperationMode(CipherOperationMode.DECRYPT);
         cipherInformation.setInType(EncodingType.BASE64); //TODO
-        DecryptionProvider baseCipher =
-                CipherFactory.createCipher(cipherInformation, keyStoreWrapper);
+        DecryptionProvider baseCipher;
+        if (symmetricEncryptionEnabled) {
+            SecretKeySpec keyStoreSecret = deriveKeyFromPassword(keyStoreWrapper.getKeyPassword());
```
Comment (hwupathum) [inline src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java:138]: Add comment

**Review by ashendes — APPROVED** [review]

File: src/main/java/org/wso2/securevault/BaseCipher.java [hunk 4]
```diff
@@ -111,22 +113,12 @@ private void init() {
             } else {
                 cipher = Cipher.getInstance(algorithm);
             }
-            if (opMode == CipherOperationMode.ENCRYPT) {
```
[the remark this answers is not in the document]
Comment (hwupathum) [reply src/main/java/org/wso2/securevault/BaseCipher.java:114]: In AES/GCM encryption, it is recommended to use separate InitialVectors for each encryption/decryption [1]. Therefore I moved the init method to encrypt and decrypt in the parent class. We have done the same implementation in carbon-crypto-service [2] too.

[1] https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/javax/crypto/Cipher.html
[2] https://github.com/wso2/carbon-crypto-service/blob/master/components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/KeyStoreBasedInternalCryptoProvider.java#L168

File: src/main/java/org/wso2/securevault/DecryptionProvider.java [hunk 5]
```diff
@@ -30,4 +32,13 @@ public interface DecryptionProvider {
...
     public byte[] decrypt(byte[] cipherText);
+
+    /**
+     * Decrypts input cipher bytes.
+     *
+     * @param cipherText    Encrypted text as a byte array.
+     * @param params        The algorithm parameters.
+     * @return Plain text as byte array.
+     */
+    public byte[] decrypt(byte[] cipherText, AlgorithmParameterSpec params);
```
[the remark this answers is not in the document]
Comment (hwupathum) [reply src/main/java/org/wso2/securevault/DecryptionProvider.java:43]: Will add a default method

File: src/main/java/org/wso2/securevault/BaseCipher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (darshanasbg) [reply src/main/java/org/wso2/securevault/BaseCipher.java:114]: Can you quote the 

> In AES/GCM encryption, it is recommended to use separate InitialVectors for each encryption/decryption [1]. 
Are we using different IVs for each encryption/decryption?

[the remark this answers is not in the document]
Comment (hwupathum) [reply src/main/java/org/wso2/securevault/BaseCipher.java:114]: > Note that GCM mode has a uniqueness requirement on IVs used in encryption with a given key. When IVs are repeated for GCM encryption, such usages are subject to forgery attacks. Thus, after each encryption operation using GCM mode, callers should re-initialize the cipher objects with GCM parameters which has a different IV value.

https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/javax/crypto/Cipher.html

[the remark this answers is not in the document]
Comment (hwupathum) [reply src/main/java/org/wso2/securevault/BaseCipher.java:114]: > Are we using different IVs for each encryption/decryption?

Yes

[the remark this answers is not in the document]
Comment (darshanasbg) [reply src/main/java/org/wso2/securevault/BaseCipher.java:114]: ack

**Review by darshanasbg — DISMISSED** [review]

File: src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java [hunk 6]
```diff
@@ -131,12 +154,43 @@ public void init(Properties properties, String id) {
                 continue;
             }
 
-            String decryptedText = new String(baseCipher.decrypt(encryptedText.trim().getBytes()));
+            String decryptedText;
+            if (DEFAULT_SYMMETRIC_ALGORITHM.equals(algorithm)) {
```
Comment (hwupathum) [inline src/main/java/org/wso2/securevault/secret/repository/FileBaseSecretRepository.java:158]: Add a comment saying this part is for GCM mode only

**Review by ashendes — APPROVED** [review]

**Pull request merged**